mshadow/extension/transpose.h

```diff
@@ -0,0 +1,90 @@
+/*!
+ *  Copyright (c) 2016 by Contributors
+ * \file transpose.h
+ * \brief support for transpose
+ * \author Junyuan Xie
+ */
+#ifndef MSHADOW_EXTENSION_TRANSPOSE_H_
+#define MSHADOW_EXTENSION_TRANSPOSE_H_
+#include <algorithm>
+#include "../extension.h"
+namespace mshadow {
+namespace expr {
+/*!
+ * \brief transpose axes of a tensor
+ * input: Tensor<Device,dim>: ishape
+ * output: Tensor<Device,dimdst> oshape[a1],oshape[a2] = ishape[a2],oshape[a1]
+ *
+ * \tparam SrcExp type of source expression
+ * \tparam DType the type of elements 
+ * \tparam dimsrc source dimension, assert a1 > a2
+ * \tparam m_a1 one dimension to be swapped, encoded by dimsrc - a1 
+ * \tparam a2 second dimension to be swapped, encoded by a2
+ */
+template<typename SrcExp, typename DType, int dimsrc>
+struct TransposeExExp:
```

Comment:
This is exactly the same with numpy. transpose
On May 31, 2016 9:35 AM, "Tianqi Chen" <notifications@github.com> wrote:

> In mshadow/extension/transpose.h
> <https://github.com/dmlc/mshadow/pull/122#discussion_r65217919>:
>
> > +#include "../extension.h"
> > +namespace mshadow {
> > +namespace expr {
> > +/*!
> > + * \brief transpose axes of a tensor
> > + * input: Tensor<Device,dim>: ishape
> > + * output: Tensor<Device,dimdst> oshape[a1],oshape[a2] = ishape[a2],oshape[a1]
> > + *
> > + * \tparam SrcExp type of source expression
> > + * \tparam DType the type of elements
> > + * \tparam dimsrc source dimension, assert a1 > a2
> > + * \tparam m_a1 one dimension to be swapped, encoded by dimsrc - a1
> > + * \tparam a2 second dimension to be swapped, encoded by a2
> > + */
> > +template<typename SrcExp, typename DType, int dimsrc>
> > +struct TransposeExExp:
>
> the name transpose as a bit confusing, consider to use a better name for
> this
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/dmlc/mshadow/pull/122/files/f374cbe8362f6a61877d7a48a4a7923da9cc37cc#r65217919>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAiudF4F1cH5Z-lLqtB91hXiKPT82t7Xks5qHGNBgaJpZM4Io6ou>
> .
>